# Pytorch Tutorial

Pytorch is a popular deep learning framework and it's easy to get started.

In [1]:
import torch
import torch.nn as nn
from torchvision import datasets, transforms
import torch.utils.data 
NUM_EPOCHS = 10
BATCH_SIZE = 128

First, we read the mnist data, preprocess them and encapsulate them into dataloader form.

In [2]:
train_dataset = datasets.MNIST(root='./mnist_dataset/',
                               train=True,
                               transform=transforms.transforms.ToTensor(),
                               download=True)
test_dataset = datasets.MNIST(root='./mnist_dataset/',
                              train=False,
                              transform=transforms.transforms.ToTensor(),
                              download=True)
batch_size = 64
train_loader = torch.utils.data.DataLoader(dataset=train_dataset,
                                           batch_size=BATCH_SIZE,
                                           shuffle=True)
test_loader = torch.utils.data.DataLoader(dataset=test_dataset,
                                          batch_size=1,
                                          shuffle=True)


Then, we define the model, object function and optimizer that we use to classify.

In [3]:
class LeNet(nn.Module):
    def __init__(self):
        super(LeNet, self).__init__()
        self.conv1 = nn.Sequential(nn.Conv2d(1, 6, 3, 1, 2),
                                   nn.ReLU(),
                                   nn.MaxPool2d(2, 2)
                                   )
        self.conv2 = nn.Sequential(nn.Conv2d(6, 16, 5),
                                   nn.ReLU(),
                                   nn.MaxPool2d(2, 2))
        self.fc1 = nn.Sequential(nn.Linear(16 * 5 * 5, 120),
                                 nn.BatchNorm1d(120),
                                 nn.ReLU())
        self.fc2 = nn.Sequential(nn.Linear(120, 120),
                                 nn.BatchNorm1d(120),
                                 nn.ReLU(),
                                 nn.Linear(120, 10))

    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = x.view(x.size()[0], -1)
        x = self.fc1(x)
        x = self.fc2(x)
        return x


net = LeNet()
my_loss = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(net.parameters(), lr=0.001)


Next, we can start to train and evaluate!

In [4]:
for epoch in range(NUM_EPOCHS):
    i = 0
    for data in train_loader:
        inputs, labels = data
        optimizer.zero_grad()
        outputs = net(inputs)
        loss = my_loss(outputs, labels)
        loss.backward()
        optimizer.step()
        i = i + 1
        if i % 100 == 0:
            print(i, "\t", "loss:",loss.item())

net.eval()
total = 0
prt = 0
for data in test_loader:
    test_input, test_label = data
    test_output = net(test_input)
    total += 1
    if torch.argmax(test_output, dim=1) == test_label:
        prt += 1
print("accuracy=\t", prt / total)



100 	 loss: 0.14536787569522858
200 	 loss: 0.059743043035268784
300 	 loss: 0.06675297021865845
400 	 loss: 0.05610867962241173
100 	 loss: 0.07410240173339844
200 	 loss: 0.03242039307951927
300 	 loss: 0.014594472013413906
400 	 loss: 0.04925417900085449
100 	 loss: 0.09165441244840622
200 	 loss: 0.01892494410276413
300 	 loss: 0.07361435145139694
400 	 loss: 0.04592376574873924
100 	 loss: 0.027191154658794403
200 	 loss: 0.011142151430249214
300 	 loss: 0.0062614199705421925
400 	 loss: 0.006285541225224733
100 	 loss: 0.03566594049334526
200 	 loss: 0.06140012666583061
300 	 loss: 0.0022845789790153503
400 	 loss: 0.012490279041230679
100 	 loss: 0.02831631898880005
200 	 loss: 0.012542365118861198
300 	 loss: 0.008582250215113163
400 	 loss: 0.0025901501066982746
100 	 loss: 0.012296225875616074
200 	 loss: 0.007965879514813423
300 	 loss: 0.025969916954636574
400 	 loss: 0.029760152101516724
100 	 loss: 0.011557401157915592
200 	 loss: 0.030154064297676086
300 	 loss: 0.005325

#### Q5:
Please print the training and testing accuracy：  
100 	 loss: 0.14536787569522858  
200 	 loss: 0.059743043035268784  
300 	 loss: 0.06675297021865845  
400 	 loss: 0.05610867962241173  
100 	 loss: 0.07410240173339844  
200 	 loss: 0.03242039307951927  
300 	 loss: 0.014594472013413906  
400 	 loss: 0.04925417900085449  
100 	 loss: 0.09165441244840622  
200 	 loss: 0.01892494410276413  
300 	 loss: 0.07361435145139694  
400 	 loss: 0.04592376574873924  
100 	 loss: 0.027191154658794403  
200 	 loss: 0.011142151430249214  
300 	 loss: 0.0062614199705421925  
400 	 loss: 0.006285541225224733  
100 	 loss: 0.03566594049334526  
200 	 loss: 0.06140012666583061  
300 	 loss: 0.0022845789790153503  
400 	 loss: 0.012490279041230679  
100 	 loss: 0.02831631898880005  
200 	 loss: 0.012542365118861198  
300 	 loss: 0.008582250215113163  
400 	 loss: 0.0025901501066982746  
100 	 loss: 0.012296225875616074  
200 	 loss: 0.007965879514813423  
300 	 loss: 0.025969916954636574  
400 	 loss: 0.029760152101516724  
100 	 loss: 0.011557401157915592  
200 	 loss: 0.030154064297676086  
300 	 loss: 0.005325149279087782  
400 	 loss: 0.07809070497751236  
100 	 loss: 0.0026013674214482307  
200 	 loss: 0.008759486488997936  
300 	 loss: 0.02036881074309349  
400 	 loss: 0.0030407439917325974  
100 	 loss: 0.002072123344987631  
200 	 loss: 0.00845141801983118  
300 	 loss: 0.02321411482989788  
400 	 loss: 0.029385928064584732  
accuracy=	 0.989